doc/cpp_domdec.rst

```diff
@@ -0,0 +1,186 @@
+
+        The number of GPUs available.
+
+        By default it is assumed that there are no GPUs.
+
+Load Balancers
+--------------
+
+Load balancing generates a :cpp:class:`domain_decomposition` given a :cpp:class:`recipe`
+and a description of the hardware on which the model will run. Currently Arbor provides
+one load balancer, :cpp:func:`partition_load_balance`, and more will be added over time.
+
+If the model is distributed with MPI, the partitioning algorithm for cells is
+distributed with MPI communication. The returned :cpp:class:`domain_decomposition`
+describes the cell groups on the local MPI rank.
+
+.. Note::
+    Power users know their model and the hardware resources that
+    they are running on very well, and know best how to partition their model
+    workload to get optimal performance.
+    The :cpp:class:`domain_decomposition` type is kept very simple, and is
+    independent of any load balancing algorithm, so power users can supply their
+    own domain decomposition without using one of the built-in load balancers.
+
+.. cpp:namespace:: arb
+
+.. cpp:function:: domain_decomposition partition_load_balance(const recipe& rec, hw::node_info nd)
+
+    Construct a :cpp:class:`domain_decomposition` that distributes the cells
+    in the model described by :cpp:var:`rec` over the hardware resources described
+    by `hw::node_info`.
+
+Decomposition
+-------------
+
+Documentation for the data structures used to describe domain decompositions.
+
+.. cpp:namespace:: arb
+
+.. cpp:enum-class:: backend_kind
+
+    Used to indicate which hardware backend to use for running a :cpp:class:`cell_group`.
+
+    .. cpp:enumerator:: multicore
+
+        Use multicore backend.
+
+    .. cpp:enumerator:: gpu
+
+        Use gpu back end.
+
+        .. Note::
+            Can only be set if the :cpp:class:`cell_group` type supports the GPU backend.
+
+.. cpp:class:: domain_decomposition
+
+    Describes a domain decomposition. Solely responsible for describing the
+    distribution of cells across cell groups and domains.
+    Holds cell group descriptions (:cpp:member:`groups`) for cells assigned to
+    the local domain, and a helper function (:cpp:member:`gid_domain`) used to
+    look up which domain a cell has been assigned to.
+    Also has meta-data about the number of cells in the global model, and the
+    number of domains over which the model is destributed.
+
+    A load balancing algorithm generates the domain decomposition, for example see
+    :cpp:func:`partition_load_balance`.
+
+    .. cpp:function:: bool is_local_gid(cell_gid_type gid) const
+
+        Tests whether the cell with :cpp:var:`gid` is on the local domain.
+
+    .. cpp:member:: std::function<int(cell_gid_type)> gid_domain
+
+        A callback function that returns the domain id that a cell assigned to
+        (using global identifier :cpp:var:`gid`).
+        Set by the load balancing algorithm that generates the domain decomposition.
```

Comment:
While technically a callback, I think it is better couched in terms of supplying a way of querying the (implicit) mapping from cell gids to domains: we don't specify when or how it would be called, and it should be a pure function in that it shouldn't have side-effects and the result should depend solely on the gid parameter.

I don't think there is any need to state that this is supplied by the load balancing algorithm: if the object comes from that, then of course it is, but we also want to make it clear that users are not tied to using that balancer.

Reply:
Good point.